[WIP] Reduce friction with Drupal 9 development on Pantheon

greg-1-anderson opened the pull request:
This PR suggests a number of changes that could be made to reduce friction commonly encountered when doing Drupal 9 development on Pantheon using Integrated Composer.

- .gitignore: If Drupal introduces a new scaffold file (or renames an existing scaffold file), then the .gitignore for all Drupal 9 projects on Pantheon will no longer be in sync with the list of files placed by Composer / Drupal Scaffold. This causes a build failure. This PR changes the default .gitignore to by default ignore *all* locations, and adds exclusions for locations that should be committed instead, e.g. `/web/modules/custom`.
- `"pantheon-upstreams/upstream-configuration": "self.version"`: Still investigating, I think this should go back to `*`.
- Move dependencies to top-level composer.json:
- `/contrib/` instead of `/composer/`:

markfullmer commented:
Thanks for proposing these changes, Greg, and for succinctly summarizing the rationale. We at UT Austin developed an Integrated Composer Pantheon upstream modeled on `pantheon-upstreams/drupal-project`, which so far has seemed to work well. That upstream repository can be seen at [utdk/utdk-project](https://github.com/utdk/utdk-project) and our architectural rationale is summarized in its [wiki](https://github.com/utdk/utdk-project/wiki).

Below is our take on how/whether the proposed changes would apply to our upstream.

### Change the default .gitignore to by default ignore all locations

We experienced the pain point of the failed build process and understand how this change would reduce that friction, but we probably wouldn't adopt this in our Pantheon IC upstream. Rationale:
- Most of our users aren't experts at Git, and understanding the double-negative of exclusions in a `.gitignore` file would be a hard sell
- This would put the onus of work on the site developer to explicitly take action when adding non-Composer-based elements to the codebase
- This change would increase the possibility of merge conflicts between our upstream repo's `.gitignore` and an individual site's `.gitignore`
- We already have a release packaging step that checks for files added that are not .gitignored, so we've mostly mitigated the original pain point in our upstream packaging

### Set "minimum-stability": "stable"

We follow the reasoning here, and would likely adopt this change, but have one question: for our internal development of Composer packages that we provide to our users, we need to require `dev` versions (Git branches), both in local development and in our automated testing processes. Do you have a recommendation on how to accommodate this (e.g., add `composer config minimum-stability dev` during development, use inline aliases (e.g., `composer require <vendor>/<name> "dev-branchname as 1.0.0"`)?

### Move dependencies to top-level composer.json

This change probably doesn't apply to how we've architected our Pantheon IC upstream. We decided to set the version of `drupal/core-recommended` and other packages in a Drupal installation profile, provided as a separate Composer package, and we decided to set an explicit version constraint rather than a floating version constraint. By doing so, we can more tightly control, via the Pantheon upstream, when a Drupal core update is available. This also means that an upstream code change is involved in incrementing versions. For this reason, we think we would want to keep this in `upstream-configuration/` in order to reduce the likelihood of Git merge conflicts in site repositories.

Additionally, we haven't found that it is "very hard to alter or move" requirements found in `upstream-configuration`; we typically get around this by using Composer inline aliases to bypass the requirement set in `upstream-configuration`.

### /contrib/ instead of /composer/

We already adopted this decision, more out of that seeming to be the convention (e.g., [drupal-project/drupal-composer](https://github.com/drupal-composer/drupal-project/blob/9.x/composer.json#L70-L72)). We also adopted the convention of `web/modules/custom` for the Composer installer type `drupal-custom-module` (as described in [#13 of our architecture decisions](https://github.com/utdk/utdk-project/wiki/Architecture-decisions)). So, both `web/modules/contrib` and `web/modules/custom` are reserved for Composer packages, while Git-tracked modules go in `web/modules/<site-namespace>`.

greg-1-anderson commented:
@markfullmer Thank you for your feedback on these proposals.

> Most of our users aren't experts at Git, and understanding the double-negative of exclusions in a .gitignore file would be a hard sell

Thank you for raising this concern; others have said similar things. Note, however, that your current .gitignore file already includes a couple of `!` rules. If these are not causing too much trouble, perhaps you might consider using a simpler solution than the one originally proposed, even if it still relies on a couple of `!` expressions? Please take a look at the new diff for this PR. I considered your feedback and reworked the .gitignore to be as similar to the existing one as possible. With the new .gitignore file, you **would** need to add more `!`-double-negatives for any file you commit directly at the document root (/web/*), but you would **not** need to add any such rules for files committed in other locations inside the document root (e.g. /web/sites/default/*), nor do you need to add any ignore rules for scaffold files placed inside of the document root. These rules are particularly convenient if you scaffold your favicon files & c.

> or our internal development of Composer packages that we provide to our users, we need to require dev versions (Git branches), both in local development and in our automated testing processes. Do you have a recommendation on how to accommodate this?

Yes. It is not necessary to use a version alias for this. Dependencies **required directly in the root-level composer.json** may override the minimum stability setting if they explicitly request a less-stable version. So, for example, you could use `composer require drupal/module:dev-featurebranch`, or `composer require drupal/other-module:^2@beta`. Dependencies in other projects, e.g. upstream-configuration/composer.json, however, must respect the minimum stability set at the project root. If your distribution wishes to deploy a beta version of a module to all sites that use it, then you'll need to use minimum-stability: beta in your top-level composer.json file. This is still far preferable to using minimum-stability: dev, which has all sorts of unfortunate edge-case behavior. For this reason, we recommend that you not put dev versions of modules in your upstream configuration file. If you must do this, then you must use minimum-stability: dev, and accept any problems that causes.

> We decided to set the version of drupal/core-recommended and other packages in a Drupal installation profile, provided as a separate Composer package

This is a sound strategy. The change to move dependencies to the top-level composer.json file does not have very much impact on custom upstreams, which are free to change both the initial top-level composer.json file and also the upstream-configuration/composer.json file distributed with pantheon-upstreams/drupal-project. This alteration will mainly benefit individual sites that use the standard Pantheon upstream directly.

> while Git-tracked modules go in web/modules/<site-namespace>

This is a good idea. I added a comment to the .gitignore file suggesting it.

Please let me know if you have any further questions or feedback on the proposal.

greg-1-anderson commented:
I simplified even more in  6fcc846

stovak reviewed (approved):
It has a great beat and I really like dancing to it. Two thumbs way up 👍🏻👍🏻

uberhacker commented:
For those wishing to include the changes in this PR to their own project, if it helps any, I executed the following:
```bash
$ git clone -b refactor-project git@github.com:pantheon-systems/drupal-project.git
$ cd drupal-project
$ cp .gitignore composer.json /path/to/pantheon/project
$ cd upstream-configuration
$ cp .gitignore composer.json /path/to/pantheon/project/upstream-configuration
```
Replace `/path/to/pantheon/project` with the path to your new Pantheon project.

mbaynton reviewed (approved):
This PR has been in my review requests list on GitHub for years and I don't seem to have any other ways to clear it.
